### PR TITLE
Added the capability to specify time interval and removed function to…

### LINKED ISSUE #1829: [TASK] Update DSS Postprocessor to accept user designated time intervals
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
For the current version, the user is restricted to postprocess the full time interval provided. If the user only intends to postprocess a portion of the data to investigate local temporal behavior, a version with only the specified pivot parameter, feature, and target must be provided.

**Describe the solution you'd like**
The solution should allow the user to specify a time interval as an option.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by yoshiurr-INL**
… postprocess more than one feature and target set per step

--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
close #1829 

##### What are the significant changes in functionality due to this change request?
Allows users to specify time intervals in ratio terms of the full time interval (pivot parameter). To prevent confusion, the capability to postprocess multiple pairs of features and targets in one single step has been deprecated. To postprocess multiple time intervals, the user will now have to create a new step and add it to the sequence.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Review by wangcj05 — COMMENTED** [review]
I would like to keep the capacities to compute multiple feature-target pairs in the single post-processors. @yoshiurr-INL Please let me know what do you think.

**Comment by yoshiurr-INL** [comment]
I considered reenabling the multiple feature-target pairs but realized this would essentially require the user to create multiple DataObjects and OutStreams anyway due to potentially using different pivot parameters depending on how many data points is provided from the feature and target.

If we recall, 'pivotParameterFeature' and 'pivotParameterTarget' are utilized for this postprocessor. If the grids are nonuniform, it is possible for one transient to use 'pivotParameterFeature' and for another transient to use 'pivotParameterTarget' due to the auto-selection algorithm implemented to choose coarser grids and have the finer data interpolate/extrapolate.

Then, the pivot parameter for the output DataObject cannot be described using a single pivot parameter. To avoid this risk, I propose to keep the current changes.

For the suggestion to add an attribute to provide the user the choice to use ratio or raw time values, I agree that should be an option and will make changes ASAP.

**Comment by yoshiurr-INL** [comment]
The capability to use ratios and raw values is now available.

File: doc/user_manual/PostProcessors/Validation.tex [hunk 1]
```diff
@@ -105,7 +108,19 @@ \subsubsection{Validation PostProcessors}
...
     \item \xmlNode{pivotParameterFeature}, \xmlDesc{string, required field}, specifies the pivotParameter for a feature <HistorySet>. The feature pivot parameter is the shared index of the output variables in the data object.
     \item \xmlNode{pivotParameterTarget}, \xmlDesc{string, required field}, specifies the pivotParameter for a target <HistorySet>. The target pivot parameter is the shared index of the output variables in the data object.
-    \item \xmlNode{multiOutput}, \xmlDesc{string, required field}, to extract raw values for the HistorySet. The user must use ‘raw values’ for the full set of metrics’ calculations to be dumped.
+    \item \xmlNode{separateFeatureData}, \xmlDesc{string, optional field}, specifies the custom feature interval to apply DSS postprocessing. The string should contain three parts; start time, `|', and end time all in one. For example, 0.0|0.5.
+      The start and end time should be in ratios or raw values of the full interval. In this case 0.5 would be either the midpoint time or time 0.5 of the given time units. This is node is not required and if not provided, the default is the full time interval.
```
Comment (wangcj05) [inline doc/user_manual/PostProcessors/Validation.tex:112]: change "This is node is" to "This node is"

File: doc/user_manual/PostProcessors/Validation.tex [hunk 2]
```diff
@@ -105,7 +108,19 @@ \subsubsection{Validation PostProcessors}
...
+      the following attributes need to be specified:
+      \begin{itemize}
+        \item \xmlAttr{type}, \xmlDesc{required string attribute}, options are `ratio' or `raw\_values'. The default is `ratio'.
```
Comment (wangcj05) [inline doc/user_manual/PostProcessors/Validation.tex:115]: change "required" to "optional"

File: doc/user_manual/PostProcessors/Validation.tex [hunk 3]
```diff
@@ -105,7 +108,19 @@ \subsubsection{Validation PostProcessors}
...
+      \end{itemize}
+    \item \xmlNode{separateTargetData}, \xmlDesc{string, optional field}, specifies the custom target interval to apply DSS postprocessing. The string should contain three parts; start time, `|', and end time all in one. For example, 0.0|0.5.
+      The start and end time should be in ratios or raw values of the full interval. In this case 0.5 would be either the midpoint time or time 0.5 of the given time units. This is node is not required and if not provided, the default is the full time interval.
```
Comment (wangcj05) [inline doc/user_manual/PostProcessors/Validation.tex:118]: change "This is node is" to "This node is"

File: doc/user_manual/PostProcessors/Validation.tex [hunk 4]
```diff
@@ -105,7 +108,19 @@ \subsubsection{Validation PostProcessors}
...
+      the following attributes need to be specified:
+      \begin{itemize}
+        \item \xmlAttr{type}, \xmlDesc{required string attribute}, options are `ratio' or `raw\_values'. The default is `ratio'.
```
Comment (wangcj05) [inline doc/user_manual/PostProcessors/Validation.tex:121]: change "required" to "optional"

File: doc/user_manual/PostProcessors/Validation.tex [hunk 5]
```diff
@@ -105,7 +108,19 @@ \subsubsection{Validation PostProcessors}
...
+      \end{itemize}
+    \item \xmlNode{multiOutput}, \xmlDesc{string, required field}, to extract raw values for the HistorySet. The user must use `raw values' for the full set of metrics’ calculations to be dumped.
```
Comment (wangcj05) [inline doc/user_manual/PostProcessors/Validation.tex:123]: Do you have plan to add different options for "multiOutput"? If not, it seems you do not need to expose this node to user, you can directly define the value for multiOutput in the init method of the DSS class.

File: doc/user_manual/PostProcessors/Validation.tex [hunk 6]
```diff
@@ -133,18 +146,71 @@ \subsubsection{Validation PostProcessors}
...
+      <scaleOmega>1</scaleOmega>
     </PostProcessor>
     ...
   <Models>
   ...
+  <DataObjects>
+    ...
+    <HistorySet name="pp1_out">
+      <Output>
+          dss_x2_x1,total_distance_x2_x1,process_time_x2_x1,standard_deviation_x2_x1
```
Comment (wangcj05) [inline doc/user_manual/PostProcessors/Validation.tex:190]: Could you document the outputs of the PPDSS in the user manual? The information that you can get from the PPDSS, and the way to retrieve the information.

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py [hunk 7]
```diff
@@ -57,14 +57,32 @@ class cls.
...
     inputSpecification.addSub(pivotParameterTargetInput)
+    separateFeatureDataInput = InputData.parameterInputFactory("separateFeatureData", contentType=InputTypes.StringType,
+                                                                descr="""Time points to separate feature data and conduct DSS postprocessing
+                                                                on selected time interval. Values must be fractions of the full time interval.
+                                                                To distinguish strart and end, '|' should be placed between both values. Example: 0.0|0.4
+                                                                To provide more than one selection for the same data set,
+                                                                '+' should be placed at the end of each interval to start defining the new interval. Example: 0.0|0.4+0.4|1.0
+                                                                If 'None' is provided, the DSS postprocessing will be applied to the full interval.""")
+    separateFeatureDataInput.addParam("type", InputTypes.StringType)
```
Comment (wangcj05) [inline ravenframework/Models/PostProcessors/Validations/PPDSS.py:67]: please add description for "type" using "descr"

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py [hunk 8]
```diff
@@ -57,14 +57,32 @@ class cls.
...
+    inputSpecification.addSub(separateFeatureDataInput)
+    separateTargetDataInput = InputData.parameterInputFactory("separateTargetData", contentType=InputTypes.StringType,
+                                                                descr="""Time points to separate target data and conduct DSS postprocessing
+                                                                on selected time interval. Values must be fractions of the full time interval.
+                                                                To distinguish strart and end, '|' should be placed between both values. Example: 0.0|0.4
+                                                                To provide more than one selection for the same data set,
+                                                                '+' should be placed at the end of each interval to start defining the new interval. Example: 0.0|0.4+0.4|1.0
+                                                                If 'None' is provided, the DSS postprocessing will be applied to the full interval.""")
+    separateTargetDataInput.addParam("type", InputTypes.StringType)
```
Comment (wangcj05) [inline ravenframework/Models/PostProcessors/Validations/PPDSS.py:76]: please add description for "type" using "descr"

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py [hunk 9]
```diff
@@ -114,6 +136,18 @@ def _handleInput(self, paramInput):
         self.pivotParameterFeature = child.value
       elif child.getName() == 'pivotParameterTarget':
         self.pivotParameterTarget = child.value
+      elif child.getName() == 'separateFeatureData':
+        if 'type' not in child.parameterValues.keys():
+          self.raiseAnError(IOError, 'Tag separateFeatureData must have attribute "type"')
```
Comment (wangcj05) [inline ravenframework/Models/PostProcessors/Validations/PPDSS.py:141]: In the manual, you mentioned, there is a default value for "type", please check it and make it consistent.

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py [hunk 10]
```diff
@@ -114,6 +136,18 @@ def _handleInput(self, paramInput):
...
+        else:
+          self.separateFeatureType = child.parameterValues["type"]
+          self.separateFeatureData = child.value
+      elif child.getName() == 'separateTargetData':
+        if 'type' not in child.parameterValues.keys():
+          self.raiseAnError(IOError, 'Tag separateTargetData must have attribute "type"')
```
Comment (wangcj05) [inline ravenframework/Models/PostProcessors/Validations/PPDSS.py:147]: see above comment

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py [hunk 11]
```diff
@@ -169,192 +198,266 @@ def _evaluate(self, datasets, **kwargs):
...
-          if abs(1-scaleRatioBeta) > 10**(-4):
-            self.raiseAnError(IOError, "Beta scaling ratio:",scaleRatioBeta,"must be 1")
-        elif self.scaleType == 'identity':
-          timeScalingRatio = 1
-          if abs(1-scaleRatioBeta) > 10**(-4) or abs(1-scaleRatioOmega) > 10**(-4):
-            self.raiseAnError(IOError, "Either beta or omega scaling ratio are not 1. Both must be 1")
-        else:
-          self.raiseAnError(IOError, "Scaling Type",self.scaleType, "is not provided")
+    if len(self.features) > 1 or len(self.targets) > 1:
+      self.raiseAnError(IOError, "The number of inputs for features or targets is greater than 1. Please restrict to one set per step.")
```
Comment (wangcj05) [inline ravenframework/Models/PostProcessors/Validations/PPDSS.py:202]: In the manual, you mentioned the extra features/targets will be ignore, here an error will be raised. Please update the manual to reflect this error.

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py [hunk 12]
```diff
@@ -169,192 +198,266 @@ def _evaluate(self, datasets, **kwargs):
...
+      outputDict['total_distance_'+nameTarg[1]+'_'+nameFeat[1]] = distanceTotal[cnt]
+      outputDict['feature_beta_'+nameTarg[1]+'_'+nameFeat[1]] = featureBeta[cnt]
+      outputDict['target_beta_'+nameTarg[1]+'_'+nameFeat[1]] = targetBeta[cnt]
+      outputDict['feature_omega_'+nameTarg[1]+'_'+nameFeat[1]] = featureOmegaNormScaled[cnt]
+      outputDict['target_omega_'+nameTarg[1]+'_'+nameFeat[1]] = targetOmegaNorm[cnt]
+      outputDict['feature_D_'+nameTarg[1]+'_'+nameFeat[1]] = featureD[cnt]
+      outputDict['target_D_'+nameTarg[1]+'_'+nameFeat[1]] = targetD[cnt]
+      outputDict['process_time_'+nameTarg[1]+'_'+nameFeat[1]] = newfeatureData[1][cnt]
+      outputDict['standard_deviation_'+nameTarg[1]+'_'+nameFeat[1]] = sigma[cnt]
+      rlz.append(outputDict)
```
Comment (wangcj05) [inline ravenframework/Models/PostProcessors/Validations/PPDSS.py:459]: please document these outputs in the manual

File: tests/framework/PostProcessors/Validation/test_validation_dss.xml [hunk 13]
```diff
@@ -3,19 +3,19 @@
   <TestInfo>
     <name>framework/PostProcessors/Validation/test_validation_dss</name>
     <author>yoshrk</author>
-    <created>2021-03-16</created>
+    <created>2022-05-13</created>
```
Comment (wangcj05) [inline tests/framework/PostProcessors/Validation/test_validation_dss.xml:6]: You may not need to change the date, but add additional <revision> node in <revisions>

**Review by wangcj05 — CHANGES_REQUESTED** [review]
@yoshiurr-INL I have some comments for your consideration.

File: doc/user_manual/PostProcessors/Validation.tex (hunk 1 above)
Comment (yoshiurr-INL) [reply doc/user_manual/PostProcessors/Validation.tex:112]: Edited.

File: doc/user_manual/PostProcessors/Validation.tex (hunk 2 above)
Comment (yoshiurr-INL) [reply doc/user_manual/PostProcessors/Validation.tex:115]: Edited.

File: doc/user_manual/PostProcessors/Validation.tex (hunk 3 above)
Comment (yoshiurr-INL) [reply doc/user_manual/PostProcessors/Validation.tex:118]: Edited.

File: doc/user_manual/PostProcessors/Validation.tex (hunk 4 above)
Comment (yoshiurr-INL) [reply doc/user_manual/PostProcessors/Validation.tex:121]: Edited.

File: doc/user_manual/PostProcessors/Validation.tex (hunk 6 above)
Comment (yoshiurr-INL) [reply doc/user_manual/PostProcessors/Validation.tex:190]: Added all information of possible postprocessed data that can be retrieved.

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py (hunk 7 above)
Comment (yoshiurr-INL) [reply ravenframework/Models/PostProcessors/Validations/PPDSS.py:67]: Description has been added for attribute "type" in separateFeatureData and separateTargetData.

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py (hunk 8 above)
Comment (yoshiurr-INL) [reply ravenframework/Models/PostProcessors/Validations/PPDSS.py:76]: Description has been added for attribute "type" in separateFeatureData and separateTargetData.

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py (hunk 9 above)
Comment (yoshiurr-INL) [reply ravenframework/Models/PostProcessors/Validations/PPDSS.py:141]: Have removed the error message and now use the default.

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py (hunk 10 above)
Comment (yoshiurr-INL) [reply ravenframework/Models/PostProcessors/Validations/PPDSS.py:147]: Have removed the error message and now use the default.

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py (hunk 11 above)
Comment (yoshiurr-INL) [reply ravenframework/Models/PostProcessors/Validations/PPDSS.py:202]: Manual now mentions error output.

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py (hunk 12 above)
Comment (yoshiurr-INL) [reply ravenframework/Models/PostProcessors/Validations/PPDSS.py:459]: Has been documented as mentioned in previous comment.

File: tests/framework/PostProcessors/Validation/test_validation_dss.xml (hunk 13 above)
Comment (yoshiurr-INL) [reply tests/framework/PostProcessors/Validation/test_validation_dss.xml:6]: Added new revision node.

File: doc/user_manual/PostProcessors/Validation.tex [hunk 14]
```diff
@@ -89,13 +89,12 @@ \subsubsection{Validation PostProcessors}
...
-    targets. \nb Each target is paired with a feature listed in xml node \xmlNode{Features}. To enable user defined
-    time interval selection, this postprocessor will only consider the first feature name provided. If user provides more than one,
-    it will ignore the other inputs.
+  \item \xmlNode{Features}, \xmlDesc{comma separated string, required field}, specifies the names of the features. Make sure the feature data are normalized by a nominal value.
+    To enable user defined time interval selection, this postprocessor will only consider the first feature name provided. If user provides more than one,
+    it will output an error.
+  \item \xmlNode{Targets}, \xmlDesc{comma separated string, required field}, specifies the names of the targets. Make sure the feature data are normalized by a nominal value. \nb Each target is paired with a feature listed in xml node \xmlNode{Features}.
+    To enable user defined time interval selection, this postprocessor will only consider the first feature name provided. If user provides more than one,
+    itt will output an error.
```
Comment (wangcj05) [inline doc/user_manual/PostProcessors/Validation.tex:97]: itt --> it

**Review by wangcj05 — CHANGES_REQUESTED** [review]
I have some minor comments for you to consider.

File: doc/user_manual/PostProcessors/Validation.tex [hunk 15]
```diff
@@ -134,7 +132,27 @@ \subsubsection{Validation PostProcessors}
...
+      \item \textbf{target\_omega\_targetName\_featureName}, provides the normalized target first order derivative data.  `targetName' and `featureName' are the string names of the input target and feature.
+      \item \textbf{feature\_D\_targetName\_featureName}, provides the feature temporal displacement rate (second order term) data.  `targetName' and `featureName' are the string names of the input target and feature.
+      \item \textbf{target\_D\_targetName\_featureName}, provides the target temporal displacement rate (second order term) data.  `targetName' and `featureName' are the string names of the input target and feature.
+      \item \textbf{process\_time\_targetName\_featureName}, provides the shared process time data.  `targetName' and `featureName' are the string names of the input target and feature.
+      \item \textbf{standard\_error\_targetName\_featureName}, provides the standard error of the overall transient data.  `targetName' and `featureName' are the string names of the input target and feature.
+    \end{itemize}
+  \item \xmlNode{Options}
+  \begin{itemize}
+    \item \xmlNode{pivotParameter}
+  \end{itemize}
```
[the remark this answers is not in the document]
Comment (yoshiurr-INL) [reply doc/user_manual/PostProcessors/Validation.tex:153]: Do not need these. Deleted.

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py [hunk 16]
```diff
@@ -61,19 +61,17 @@ class cls.
...
                                                                 on selected time interval. Values must be fractions of the full time interval.
                                                                 To distinguish strart and end, '|' should be placed between both values. Example: 0.0|0.4
-                                                                To provide more than one selection for the same data set,
-                                                                '+' should be placed at the end of each interval to start defining the new interval. Example: 0.0|0.4+0.4|1.0
                                                                 If 'None' is provided, the DSS postprocessing will be applied to the full interval.""")
-    separateFeatureDataInput.addParam("type", InputTypes.StringType)
+    separateFeatureDataInput.addParam("type", InputTypes.StringType, descr="""Specifies what type of user defined time intervals have been selected from
+                                                                    the feature pivot parameter. Options are currently `ratio' or `raw_values'""")
```
[the remark this answers is not in the document]
Comment (yoshiurr-INL) [reply ravenframework/Models/PostProcessors/Validations/PPDSS.py:66]: Changed.

File: ravenframework/Models/PostProcessors/Validations/PPDSS.py [hunk 17]
```diff
@@ -61,19 +61,17 @@ class cls.
...
                                                                 on selected time interval. Values must be fractions of the full time interval.
                                                                 To distinguish strart and end, '|' should be placed between both values. Example: 0.0|0.4
-                                                                To provide more than one selection for the same data set,
-                                                                '+' should be placed at the end of each interval to start defining the new interval. Example: 0.0|0.4+0.4|1.0
                                                                 If 'None' is provided, the DSS postprocessing will be applied to the full interval.""")
-    separateTargetDataInput.addParam("type", InputTypes.StringType)
+    separateTargetDataInput.addParam("type", InputTypes.StringType, descr="""Specifies what type of user defined time intervals have been selected from
+                                                                    the target pivot parameter. Options are currently `ratio' or `raw_values'""")
```
[the remark this answers is not in the document]
Comment (yoshiurr-INL) [reply ravenframework/Models/PostProcessors/Validations/PPDSS.py:74]: Changed.

**Review by wangcj05 — APPROVED** [review]

**Comment by wangcj05** [comment]
PR checklist is satisfied.

**Pull request merged**